SIMSBIOHUB-602: Include Animals in the Survey page map

mauberti-bc opened the pull request:
# Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-602 (for the map popups)

# Description of changes

- Added ```ajv``` dependency to fix a dependency issue (should it be in overrides?)
- Added Animals to the Survey page data

Related PR that fixes the detailed critter endpoint SQL: https://github.com/bcgov/critterbase-api/pull/60 (merged)

# Testing Notes

On Survey page, the observation, animal, and telemetry data should appear in the map and the table below. All items on the map should have popups. The sample site/study area additionally have a tooltip.

mauberti-bc committed:
wip: include animals in survey map data

mauberti-bc committed:
merge dev

mauberti-bc committed:
get detailed survey critters

mauberti-bc committed:
merge dev

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into survey-page

mauberti-bc committed:
include animals in survey data table

mauberti-bc committed:
Merge branch 'dev' into survey-page

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into survey-page

mauberti-bc committed:
Merge branch 'survey-page' of github.com:bcgov/biohubbc into survey-page

mauberti-bc committed:
fix imports

mauberti-bc committed:
make fix

mauberti-bc committed:
refactor survey spatial data map

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into survey-page

mauberti-bc committed:
fix keys in autocomplete components

mauberti-bc committed:
increase container heights on summary page

mauberti-bc committed:
wip: include animal captures in survey page map

mauberti-bc committed:
wip: include detailed critters

mauberti-bc committed:
consolidate survey data layers

mauberti-bc committed:
include additional columns in telemetry data table

mauberti-bc committed:
fix code smells

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into survey-page

NickPhura committed:
Fix some react warnings (dependencies, useCallback).
Update AsyncAutocompleteDataGridEditCell.tsx to be generic, move additional props to TaxonomyDataGridEditCell.tsx.
WIP: debugging empty response from critterbase

NickPhura committed:
Remove logs.
Update openapi spec.
Fix infinite refresh on telemetry page.

MacDeluca commented on app/src/utils/Utils.tsx:
```diff
@@ -478,3 +478,40 @@ export const getRandomHexColor = (seed: number, min = 100, max = 170): string =>
+/**
+ * Creates a GeoJSON point from a set of coordinates and associated properties.
+ *
+ * @param latitude - The latitude coordinate.
```
The jsdoc syntax looks slightly off @param {number} latitude - The latitude coordinate.

MacDeluca commented on app/src/utils/Utils.tsx:
```diff
@@ -478,3 +478,40 @@ export const getRandomHexColor = (seed: number, min = 100, max = 170): string =>
+ * @param longitude - The longitude coordinate.
+ * @returns A GeoJSON point
+ */
+export const createGeoJSONPoint = (latitude: number, longitude: number): Point => {
```
Maybe these helper functions could be moved to its own utils file. utils/spatial or utils/geoJson

MacDeluca commented on app/src/hooks/useContext.tsx:
```diff
@@ -147,6 +148,23 @@ export const useObservationsTableContext = (): IObservationsTableContext => {
 };
 
+/**
+ * Returns an instance of `IObservationsDataContext` from `ObservationsDataContext`.
```
TelemetryDataContext

MacDeluca commented on app/src/hooks/api/useSurveyApi.ts:
```diff
@@ -399,6 +400,22 @@ const useSurveyApi = (axios: AxiosInstance) => {
 
+  /**
+   * Retrieve a list of critters associated with the given survey with details taken from critterbase, including
+   * additiona information such as captures and mortality
```
additional**

MacDeluca commented on app/src/features/surveys/view/components/spatial-data/components/telemetry/table/SurveyDataTelemetryTable.tsx:
```diff
@@ -0,0 +1,157 @@
+   */
+  const tableRows: ITelemetryData[] = useMemo(() => {
+    return surveyContext.critterDeployments.map((item) => {
+      console.log(item);
```
👀

MacDeluca commented on app/src/constants/spatial.ts:
```diff
@@ -21,9 +21,10 @@ export const ALL_OF_BC_BOUNDARY: Feature = {
 };
 
 export const SURVEY_MAP_LAYER_COLOURS = {
```
This can be moved to constants/colours

MacDeluca commented on app/src/features/surveys/view/SurveyPage.tsx:
```diff
@@ -36,13 +35,10 @@ const SurveyPage: React.FC = () => {
         <Stack gap={3}>
           <TaxonomyContextProvider>
-            <SurveySpatialData />
+            {/* <SurveySpatialData /> */}
```
👀

MacDeluca commented on api/src/services/critterbase-service.ts:
```diff
@@ -384,6 +389,11 @@ export class CritterbaseService {
     return response.data;
   }
 
+  async getMultipleCrittersByIdsDetailed(critter_ids: string[]): Promise<ICritterDetailed[]> {
```
jsdoc

MacDeluca commented on app/src/features/surveys/view/components/spatial-data/components/animal/SurveyDataAnimal.tsx:
```diff
@@ -0,0 +1,110 @@
+      for (const capture of animal.captures ?? []) {
+        if (capture.capture_location?.latitude && capture.capture_location?.longitude) {
+          const point: ISurveyMapPoint = {
+            feature: {
```
createGeoJsonFeature?

MacDeluca commented on app/src/features/surveys/view/components/spatial-data/components/telemetry/SurveyDataTelemetry.tsx:
```diff
@@ -0,0 +1,150 @@
+          }) ?? []
+      );
+    },
+    []
```
Dependencies?

mauberti-bc committed:
address PR comments

mauberti-bc committed:
wip: prevent zoom when point is clicked

mauberti-bc committed:
wip: move map popup to staticlayer

NickPhura committed:
Map popups working.
ToDo: Custom icons, Telemetry

NickPhura committed:
Add custom icon support

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into survey-page

NickPhura committed:
Fix merge issues

NickPhura committed:
Fix linting/code smells

NickPhura committed:
Misc cleanup

NickPhura committed:
Fix package audit issues, remove unused package

NickPhura committed:
Fix npm warning

NickPhura committed:
Update telemetry popup

mauberti-bc committed:
fix labels in map popups

mauberti-bc committed:
fix missed label in map popups

NickPhura committed:
Update telemetry endpoint, add/update some unit tests

MacDeluca commented:
Bug fix:
1. Clicking on capture point on the map correctly loads the data first time, second click renders skeletons, third click works.

Some ideas (feel free to include whichever ones):
1. We could disable the tool tip when we select a map element (to prevent the tooltip rendering underneath). Hover to render the tooltip -> click to see details (only one rendering at a time).
![image](https://github.com/user-attachments/assets/ba7a7192-8ac5-4ee5-8392-6b4686b59468)
2. Tooltips missing for observations and animals / captures

MacDeluca commented on api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts:
```diff
@@ -0,0 +1,207 @@
+      const captures: ICaptureGeometry[] = [];
+
+      // Map over all captures to extract geometry as points
+      for (const critter of critters) {
```
Should this logic be moved to its own service method?

NickPhura replied:
It looks like this endpoint isn't ever called. So might be able to remove it entirely.

NickPhura replied:
In the meantime, I've put a TODO on the endpoint to flag that its unused, and can maybe be removed.

mauberti-bc replied:
I created `/critters/spatial` in SIMS before realizing it should be Critterbase with `/critters/spatial`, so yes, no problem removing it

MacDeluca committed:
fix: corrected some nits

NickPhura committed:
Fix animal popup skeleton loader bug

NickPhura commented:
> Bug fix:
> 
>     1. Clicking on capture point on the map correctly loads the data first time, second click renders skeletons, third click works.
> 

@MacQSL Pushed up a fix for this

Also added tooltips to the observation/animal/telemetry points.

NickPhura committed:
Add tooltip to map points

NickPhura committed:
Add todo for possibly deprecated endpoint

NickPhura committed:
Merge branch 'dev' into survey-page

MacDeluca committed:
fix: math.random() -> uuid(v4)

MacDeluca committed:
fix: dropping dead line

MacDeluca reviewed (dismissed):
LGTM

NickPhura committed:
ignore-skip

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into survey-page

NickPhura committed:
Fix critter interface merge issue between sims create capture vs import csv capture.
Remove unused endpoint and related types.
Fix minor bug in multer logic where a request that did not provide a media file would result in an undefined req.body error

NickPhura committed:
Fix map skeleton loading css

NickPhura reviewed (approved):
🌵
